warehouse/macaroons/caveats/_core.py

```diff
@@ -111,7 +111,13 @@ def lookup(self, /, tag: int) -> type[Caveat] | None:
 _caveat_registry = _CaveatRegistry()
 
 
-def as_caveat(*, tag: int) -> Callable[[type[T]], type[T]]:
+# TODO: The return signature detected is `"Union[Type[Dataclass], DataclassProxy]"`,
+# but the expectation is `Type[Dataclass]`.
+# See https://github.com/pydantic/pydantic/issues/4498 but not exactly the same.
+# This might not be corrected in pydantic until 2.0.
+# Original signature with type hints:
+# def as_caveat(*, tag: int) -> Callable[[type[T]], type[T]]:
+def as_caveat(*, tag: int):
```

Comment:
@dstufft Maybe you can help here? I got lost down a rabbit hole before giving up and writing this comment.